### ISSUE TITLE
Using KEEP to collect a GET-WORD! in Parse crashes the interpreter

### ISSUE THREAD

**Issue by rgchris** [metaeducation/ren-c#939]
When using KEEP, trying to collect a GET-WORD! will crash the interpreter as follows:

```
>> parse "a" [collect thing [foo: "a" keep :foo]]
Assertion failed: (P_POS >= pos_before), function N_subparse, file ../src/core/u-parse.c, line 1708.
Abort trap: 6
```

**Comment by hostilefork** [comment]
What this should do presumably parallels what COPY should do:

https://github.com/rebol/rebol-issues/issues/2363

If you'd like a light and easy chance to take your PARSE-Czar duties to the next level...this would be a good issue to try.  I think if you look at the code surrounding the assert it's not so hard to follow.  All you have to remember is that the index numbers in C are 0-based...

https://github.com/metaeducation/ren-c/blob/3e5c1acc76cb650540739c0bab59966bf783d800/src/core/u-parse.c#L1706

**Comment by hostilefork** [comment]
Old PARSE (which I am calling "PARSE3") is legacy, so I'm not worrying about edge cases like this anymore for that... 🪦

*But* UPARSE wants to be as tough as it can, and [I've been grabbing every test I can from the historical issue database as well as from Red](https://forum.rebol.info/t/dead-red-redemption-starting-uparse-on-the-right-foot/1649).  So let's try the UPARSE equivalent and see what happens.

    >> thing: <before>

    >> uparse "a" [thing: collect [foo: <here>, "a", keep seek (foo)]]
    ; null

    >> foo
    == "a"

    >> thing
    == ["a"]

**That seems right!**

* You seek'd back to the start before the "a" and then didn't consume it, so the overall parse failed (returned null)

* ...but the writing of THING to COLLECT happened because it was executed synchronously before the parse was finished.  Hence it has a block of one kept item.

* What got kept was the result of the SEEK, whose synthesized value was just the position it was asked to seek to.

Every expression has a synthesized result, and the structure is homogenous, so by-and-large such things "just work".  **[I've added it to the semi-impressive mountain-o-tests.](https://github.com/metaeducation/ren-c/commit/a0b87799ca2756acc6664eb783d296d772a40799)**

---

*Note: I'm now thinking that if you want the final position, you can just ask for that to be the synthesized result [by tacking `return <here>` on the end](https://forum.rebol.info/t/uparse-return-subtleties/1589/3?u=hostilefork):*

    >> uparse "a" [thing: collect [foo: <here>, "a", keep seek (foo)], return <here>]
    == "a"
